editor/reducer.js

```diff
@@ -454,6 +422,32 @@ export function preferences( state = STORE_DEFAULTS.preferences, action ) {
 				...state,
 				mode: action.mode,
 			};
+		case 'INSERT_BLOCKS':
+			// record the block usage and put the block in the recently used blocks
+			let blockUsage = state.blockUsage;
+			let recentlyUsedBlocks = [ ...state.recentlyUsedBlocks ];
+			action.blocks.forEach( ( block ) => {
+				const uses = ( blockUsage[ block.name ] || 0 ) + 1;
+				blockUsage = omit( blockUsage, block.name );
+				blockUsage[ block.name ] = uses;
+				recentlyUsedBlocks = [ block.name, ...without( recentlyUsedBlocks, block.name ) ].slice( 0, maxRecent );
+			} );
+			return {
+				...state,
+				blockUsage,
+				recentlyUsedBlocks,
+			};
+		case 'SETUP_EDITOR':
+			// initially populate the recently used blocks with blocks the user most frequently uses
+			const commonBlocks = getBlockTypes()
+				.filter( ( blockType ) => 'common' === blockType.category )
+				.map( ( blockType ) => blockType.name );
+			return {
+				...state,
+				recentlyUsedBlocks: [ ...state.recentlyUsedBlocks ]
```

Comment:
What if some recently used blocks do not exist any more, should we filter the list here or is this done somewhere else?

Reply:
Ah, good catch! I'll look at that now.

Reply:
I think the filtering should probably be done on editor setup, so we can filter any stored block information in the same step.

Reply:
don't you think the  `SETUP_EDITOR` is a good place for this (it's editor setup step) or are you thinking fo somewhere else?

What concerns me in moving this behaviour elsewhere is having to know the shape of the store outside the store.

Reply:
I thought `SETUP_EDITOR` would be the place to do it, in the `preferences` reducer, so we can filter out (or maybe migrate in future?) any block data that's in preferences, and should no longer be there.

Reply:
Fixed! Couple of filter functions in the preferences reducer that remove invalid block names from a list, or invalid block names from an object's properties.